From eed682609083b3d3b4b00293008506e3ac9d4994 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 25 Aug 2014 07:48:21 -0400
Subject: [x86] perf: Fix unknown NMI warning

Message-id: <1408953210-26343-14-git-send-email-jolsa@redhat.com>
Patchwork-id: 88782
O-Subject: [PATCH RHEL7.1 BZ1133083 013/322] perf, nmi: Fix unknown NMI warning
Bugzilla: 1133083
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1133083
https://bugzilla.redhat.com/show_bug.cgi?id=1133083

upstream
========
commit a3ef2229c94ff70998724cb64b9cb4c77db9e950
Author: Markus Metzger <markus.t.metzger@intel.com>
Date: Fri Feb 14 16:44:08 2014 -0800

description
===========
When using BTS on Core i7-4*, I get the below kernel warning.

$ perf record -c 1 -e branches:u ls
Message from syslogd@labpc1501 at Nov 11 15:49:25 ...
 kernel:[  438.317893] Uhhuh. NMI received for unknown reason 31 on CPU 2.

Message from syslogd@labpc1501 at Nov 11 15:49:25 ...
 kernel:[  438.317920] Do you have a strange power saving mode enabled?

Message from syslogd@labpc1501 at Nov 11 15:49:25 ...
 kernel:[  438.317945] Dazed and confused, but trying to continue

Make intel_pmu_handle_irq() take the full exit path when returning early.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kernel/cpu/perf_event_intel.c b/arch/x86/kernel/cpu/perf_event_intel.c
index 56356d4..1c395ee 100644
--- a/arch/x86/kernel/cpu/perf_event_intel.c
+++ b/arch/x86/kernel/cpu/perf_event_intel.c
@@ -1360,10 +1360,8 @@ static int intel_pmu_handle_irq(struct pt_regs *regs)
  intel_pmu_disable_all();
  handled = intel_pmu_drain_bts_buffer();
  status = intel_pmu_get_status();
- if (!status) {
-  intel_pmu_enable_all(0);
-  return handled;
- }
+ if (!status)
+  goto done;
 
  loops = 0;
 again:
-- 
1.7.1